src/dmd/doc.d

```diff
@@ -2009,14 +2009,183 @@ Lno:
     return i;
 }
 
+/****************************************************
+ * Remove a previously-inserted blank line macro.
+ * Params:
+ *  buf           = an OutBuffer containing the DDoc
+ *  iAt           = the index within `buf` of the start of the `$(DDOC_BLANKLINE)`
+ *                  macro. Upon function return its value is set to `0`.
+ *  i             = an index within `buf`. If `i` is after `iAt` then it gets
+ *                  reduced by the length of the removed macro.
+ */
+private void removeBlankLineMacro(OutBuffer *buf, ref size_t iAt, ref size_t i)
+{
+    if (!iAt)
+        return;
+
+    buf.remove(iAt, 17); // length of "$(DDOC_BLANKLINE)"
+    if (i > iAt)
+        i -= 17;
+    iAt = 0;
+}
+
+/****************************************************
+ * Detect the level of an ATX-style heading, e.g. `## This is a heading` would
+ * have a level of `2`.
+ * Params:
+ *  buf   = an OutBuffer containing the DDoc
+ *  i     = the index within `buf` of the first `#` character
+ * Returns:
+ *          the detected heading level from 1 to 6, or
+ *          0 if not at an ATX heading
+ */
+private int detectAtxHeadingLevel(OutBuffer *buf, size_t i)
+{
+    if (!global.params.markdown)
+        return 0;
+
+    const iHeadingStart = i;
+    size_t iSkipped = skipChars(buf, i, "#");
+    const headingLevel = cast(int) (iSkipped - iHeadingStart);
+    if (headingLevel > 6)
+        return 0;
+
+    i = skipChars(buf, iSkipped, " \t");
+    const emptyHeading = buf.data[i] == '\r' || buf.data[i] == '\n';
+
+    // require whitespace
+    if (!emptyHeading && i == iSkipped)
+        return 0;
+
+    return headingLevel;
+}
+
+/****************************************************
+ * Remove any trailing `##` suffix from an ATX-style heading.
+ * Params:
+ *  buf   = an OutBuffer containing the DDoc
+ *  i     = the index within `buf` to start looking for a suffix at
+ */
+private void removeAnyAtxHeadingSuffix(OutBuffer *buf, size_t i)
+{
+    size_t j = i;
+    size_t iSuffixStart = 0;
+    size_t iWhitespaceStart = j;
+    const slice = buf.peekSlice();
+    for (; j < slice.length; j++)
+    {
+        switch (slice[j])
+        {
+        case '#':
+            if (iWhitespaceStart && !iSuffixStart)
+                iSuffixStart = j;
+            break;
+        case ' ':
+        case '\t':
+            if (!iWhitespaceStart)
+                iWhitespaceStart = j;
+            break;
+        case '\r':
+        case '\n':
+            goto LendHeadingSuffix;
```

Comment:
`break;`

Reply:
I believe `break` would exit the switch statement but not the enclosing loop. That's why I used the `goto`.